### ISSUE TITLE
Strange behaviour with Awaitable[T]

### ISSUE THREAD

**Issue by Gobot1234** [microsoft/pyright#2791]
**Describe the bug**
When returning an Awaitable from a function and then awaiting it somewhere later, the type of T cannot be narrowed.

**To Reproduce**
Setup two functions one coroutine function and one that returns an awaitable, call the second from the first.

**Expected behavior**
The return type should be narrow-able from T to any subclass of T's bound.

**Screenshots or Code**
https://pyright-playground.decorator-factory.su/?gist_id=139c97c5ad8c58e75442f64e759945a9&filename=gistfile1.py

**VS Code extension or command-line**
Pylance - v2021.12.3-pre.1

**Additional context**
Removing all the async code here makes everything work.
This works fine with mypy: https://mypy-play.net/?mypy=latest&python=3.10&gist=139c97c5ad8c58e75442f64e759945a9

**Comment by erictraut** [comment]
The issue here is that you've defined a generic function in a way that the return type can go unsolved. Mypy is allowing you to get away with this when you assign the unsolved TypeVar to a variable of type `Box[int]`. If you remove the annotation for `my_foo`, mypy struggles with this condition (emits an error), but pyright does not.

Pyright is allowing you to assign the unsolved B@Foo value to `my_foo`, but it's then narrowing the type of `B@Foo` on assignment. Maybe it should avoid assignment-based narrowing in this case. I'll need to think more about what heuristics to use to determine when to skip narrowing.

My recommended approach here is to either redefine function `foo` so it doesn't potentially leave `B` unsolved or provide overloads that specify what happens when no argument is provided.

```python
@overload
async def foo(check: None = ...) -> Box[int]:
    ...

@overload
async def foo(check: Callable[[B], bool]) -> B:
    ...
```

**Comment by Gobot1234** [comment]
Thanks for the response, I'll try the overload approach.

Just curious as to why the non-async version allows you to narrow the type. https://pyright-playground.decorator-factory.su/?gist_id=139c97c5ad8c58e75442f64e759945a9&filename=non-async.py

**Comment by erictraut** [comment]
The difference between the async and non-async version is due to some heuristics in the bidirectional type inference logic. The async version needs to use bidirectional type inference through two calls (`run` and `foo`).

I've updated the assignment-based type narrowing logic. It already contained some heuristics about when to apply narrowing and when to avoid it. I've added an additional heuristic to avoid narrowing in the case where the source of the assignment contains an unsolved TypeVar. With this change, both the async and non-async versions work as you would expect.
